agent/connect/generate.go

```diff
@@ -33,3 +46,98 @@ func GeneratePrivateKey() (crypto.Signer, string, error) {
 
 	return pk, buf.String(), nil
 }
+
+// GenerateCA generates a new CA
+func GenerateCA(signer crypto.Signer, sn *big.Int, days int, uris []*url.URL) (string, error) {
+	keyID, err := KeyId(signer.Public())
+	if err != nil {
+		return "", err
+	}
+
+	name := fmt.Sprintf("Consul CA %d", sn)
+
+	// Create the CA cert
+	template := x509.Certificate{
+		SerialNumber: sn,
+		Subject: pkix.Name{
+			Country:       []string{"US"},
+			PostalCode:    []string{"94105"},
+			Province:      []string{"CA"},
+			Locality:      []string{"San Francisco"},
+			StreetAddress: []string{"101 Second Street"},
+			Organization:  []string{"HashiCorp Inc."},
+			CommonName:    name,
+		},
+		URIs:                  uris,
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign | x509.KeyUsageDigitalSignature,
+		IsCA:                  true,
+		NotAfter:              time.Now().AddDate(0, 0, days),
+		NotBefore:             time.Now(),
+		AuthorityKeyId:        keyID,
+		SubjectKeyId:          keyID,
+	}
+
+	bs, err := x509.CreateCertificate(
+		rand.Reader, &template, &template, signer.Public(), signer)
+	if err != nil {
+		return "", fmt.Errorf("error generating CA certificate: %s", err)
+	}
+
+	var buf bytes.Buffer
+	err = pem.Encode(&buf, &pem.Block{Type: "CERTIFICATE", Bytes: bs})
+	if err != nil {
+		return "", fmt.Errorf("error encoding private key: %s", err)
+	}
+
+	return buf.String(), nil
+}
+
+// GenerateCert generates a new certificate
+func GenerateCert(signer crypto.Signer, ca string, sn *big.Int, name string, days int, DNSNames []string, IPAddresses []net.IP, extKeyUsage []x509.ExtKeyUsage) (string, string, error) {
+	parent, err := ParseCert(ca)
+	if err != nil {
+		return "", "", err
+	}
+
+	signee, pk, err := GeneratePrivateKey()
+	if err != nil {
+		return "", "", err
+	}
+
+	keyID, err := KeyId(signee.Public())
+	if err != nil {
+		return "", "", err
+	}
+
+	template := x509.Certificate{
+		SerialNumber:          sn,
+		Subject:               pkix.Name{CommonName: name},
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:           extKeyUsage,
+		IsCA:                  false,
+		NotAfter:              time.Now().AddDate(0, 0, days),
+		NotBefore:             time.Now(),
+		SubjectKeyId:          keyID,
+		DNSNames:              DNSNames,
+		IPAddresses:           IPAddresses,
+	}
+
+	bs, err := x509.CreateCertificate(rand.Reader, &template, parent, signee.Public(), signer)
+	if err != nil {
+		return "", "", err
+	}
+
+	var buf bytes.Buffer
+	err = pem.Encode(&buf, &pem.Block{Type: "CERTIFICATE", Bytes: bs})
+	if err != nil {
+		return "", "", fmt.Errorf("error encoding private key: %s", err)
+	}
+
+	return buf.String(), pk, nil
+}
+
+func GenerateCSR() (string, error) {
+	return "", nil
+}
```

Comment:
I'm not super sure this is the right place to add these. Having a function in `agent/connect` called `GenerateCA` but that actually generates a _server_ CA that is inappropriate for use as a Connect CA seems very confusing.

Not only that but there are another set of very similar but not quite the same functions in this package for generating Test CA etc. but those are _connect_ ones and not server ones.

Can we put them in a separate lib? Maybe literally just in `lib` since they don't seem to have any dependencies outside of the std lib.

Reply:
Totally agreed, I was thinking something similar. I started out in connect space but it doesn't have any ties.